torch/_higher_order_ops/templated_attention.py

```diff
@@ -220,3 +249,401 @@ def templated_attention_fake_tensor_mode(
             batch_size, num_heads, seq_len_q, dtype=torch.float32
         )
         return torch.empty_like(query, memory_format=torch.contiguous_format), logsumexp
+
+
+# ---------------------------- Autograd Implementation ----------------------------
+def create_fw_bw_graph(score_mod, index_values, other_buffers):
+    # See Note:[HOP create fw_bw graph]
+
+    # All of these imports need to be here in order to avoid circular dependencies
+    from torch._dispatch.python import suspend_functionalization
+    from torch._functorch.aot_autograd import AOTConfig, create_joint
+    from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
+
+    from torch._subclasses.functional_tensor import disable_functional_mode
+    from torch.fx.experimental.proxy_tensor import disable_proxy_modes_tracing
+
+    dummy_aot_config = AOTConfig(
+        fw_compiler=None,  # type: ignore[arg-type]
+        bw_compiler=None,  # type: ignore[arg-type]
+        partition_fn=None,  # type: ignore[arg-type]
+        decompositions={},
+        num_params_buffers=0,
+        aot_id=0,
+        keep_inference_input_mutations=False,
+    )
+
+    with suspend_functionalization(), disable_functional_mode():
+        with disable_proxy_modes_tracing():
+
+            def _from_fun(t):
+                return torch.empty_strided(
+                    t.size(),
+                    t.stride(),
+                    device=t.device,
+                    dtype=t.dtype,
+                    requires_grad=t.requires_grad,
+                )
+
+            # If someone runs this hop under the default compiler backend ("eager")
+            # Then this path will be run with the actual user inputs. We convert them
+            # to fake tensors in order to not perform any actual compute.
+            maybe_tracing = torch._guards.TracingContext.try_get()
+            fake_mode = (
+                maybe_tracing.fake_mode
+                if maybe_tracing
+                else FakeTensorMode(allow_non_fake_inputs=True)
+            )
+
+            with fake_mode:
```

Comment:
the test_aot_dispatch test is failing because of this code with:


```Python
File "/home/drisspg/meta/pytorch/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/drisspg/meta/pytorch/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "<eval_with_key>.5", line 5, in forward
    add = new_empty + new_empty_2;  new_empty = new_empty_2 = None
  File "/home/drisspg/meta/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 924, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 1282, in dispatch
    return self._cached_dispatch_impl(func, types, args, kwargs)
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 1015, in _cached_dispatch_impl
    output = self._dispatch_impl(func, types, args, kwargs)
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 1372, in _dispatch_impl
    (flat_args, flat_arg_fake_tensors) = self.validate_and_convert_non_fake_tensors(
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 1607, in validate_and_convert_non_fake_tensors
    validated_args = [validate(a) for a in flat_args]
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 1607, in <listcomp>
    validated_args = [validate(a) for a in flat_args]
  File "/home/drisspg/meta/pytorch/torch/_subclasses/fake_tensor.py", line 1595, in validate
    raise AssertionError("Mixing fake modes NYI")
AssertionError: Mixing fake modes NYI

The above exception was the direct cause of the following exception:
```